### ISSUE TITLE
Reduce_on_plateau

### ISSUE THREAD

**Issue by jecampagne** [google-deepmind/optax#221]
Hi, 
 in the learning rate schedule list it may be useful to add a "reduce on plateau" scheduler which looks during a certain number of epochs if the loss is decreasing, and if not divide the learning rate by a certain amount. (possibly clipped by a minimal value).
In Pytorch  I experience that in turns out to be useful for some of my use-cases of optimizing CNN.

**Comment by rosshemsley** [comment]
Hey @jecampagne, thanks - this is a good suggestion - it could be a good example to introduce `inject_hyperparameters`.

We will leave this issue open on our backlog and consider working it into our next documentation push.

**Comment by adendek** [comment]
Hello all,

Do you need an open-source contributor? I'd like to help with this. 
Could you please give me some hints on how you would like to proceed with introducing *inject_hyperparameters*?

**Comment by rosshemsley** [comment]
Hey @adendek, yes, we would welcome contributions here!

The thing that's interesting about this is that the optimizer requires access to the current value of the loss (which is not seen by the current API).

One way to achieve this is to add the current loss as a parameter, and then to inject it using inject_hyperparameters.

There may be others, though.
Please, feel free to take a look and see what you can come up with on this!

The idea on introducing inject_hyperparameters was to write a colab example for the documentation, which uses a real practical example (such as this). Though this can (and probably should) be added as a separate change :)

**Comment by pharringtonp19** [comment]
@adendek I would like to help work on this implementation if you need an extra set of hands. I would be happy to discuss this offline via email if you want as well -- pharringtonp19@gmail.com

**Comment by n2cholas** [comment]
I like this suggestion a lot, @jecampagne! However, it's not clear to me how `inject_hyperparams` would help in this situation. `inject_hyperparams` is a utility that allows you to pass in a `ScheduleFn` where a scalar hyperparameter is expected when the optimizer is created. In its current form, it does not allow you to pass in arbitrary values at runtime, which is what this proposal would require.

One option is to implement a separate utility that updates the `opt_state` with the current loss. For example,

```py
tx = optax.chain([optax.sgd(0.1), optax.reduce_lr_on_plateau(...)])
...
opt_state = optax.update_state_with_loss(opt_state, loss_scalar)
updates, opt_state = tx.update(grads, opt_state)
```

`optax.update_state_with_loss` could check the `opt_state` tree for a `ReduceLROnPlateauState` and update it with the loss_scalar. When the `opt_state` does not contain `ReduceLROnPlateauState`, `update_state_with_loss` would do nothing.

Any thoughts on this? It somewhat breaks the `optax` model of only needing to worry about `init` and `update` regardless of gradient transformation, but this would not be unprecedented (e.g. `optax.dpsgd` required per-example gradients unlike all the other optax transformations).

**Comment by OhadRubin** [comment]
I agree, it is indeed too complicated for a user, it's possible to use this with "GradientTransformationWithExtraArgs" where the extra arg is the loss

**Comment by OhadRubin** [comment]
See https://github.com/deepmind/optax/pull/505

**Comment by vroulet** [comment]
Thanks for all the comments.
I think it has now been implemented in https://github.com/google-deepmind/optax/pull/629.
If that PR did not answer all desired needs, feel free to open a new issue.
